File: src/board.rs
use std::fmt::{self, Debug};
use std::ops::{Index, IndexMut};

use crate::coordinate::{Coordinate, CoordinateIter};

#[derive(Clone, Copy, Default)]
pub enum Ring {
    Empty,
    #[default]
    Vacant,
    Occupied(Marble),
}

impl PartialEq for Ring {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Empty, Self::Empty) => true,
            (Self::Empty, _) | (_, Self::Empty) => false,
            _ => true,
        }
    }
}

impl Debug for Ring {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Empty => write!(f, "."),
            Self::Vacant => write!(f, "O"),
            Self::Occupied(Marble::White) => write!(f, "\x1b[107;30m@\x1b[0m"),
            Self::Occupied(Marble::Gray) => write!(f, "\x1b[100;30m@\x1b[0m"),
            Self::Occupied(Marble::Black) => write!(f, "\x1b[97;40m@\x1b[0m"),
        }
    }
}

#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum Marble {
    White,
    Gray,
    Black,
}

pub struct Board {
    pub(crate) data: Vec<Ring>,
}

impl Board {
    pub fn new() -> Self {
        let data = vec![Ring::default(); 49];
        let mut output = Self { data };

        for coord in CoordinateIter::new(Coordinate::new(0, 0), Coordinate::new(6, 6), 6) {
            if coord.x > coord.y + 3 || coord.y > coord.x + 3 {
                output[coord] = Ring::Empty;
            }
        }

        output
    }

    pub fn get(&self, coord: Coordinate) -> Option<&Ring> {
        if coord.x >= 7 || coord.y >= 7 {
            return None;
        }
        self.data.get(coord.x + 7 * coord.y)
    }

    pub fn get_option(&self, coord: Option<Coordinate>) -> Option<&Ring> {
        if let Some(coord) = coord {
            self.get(coord)
        } else {
            None
        }
    }
}

impl Index<Coordinate> for Board {
    type Output = Ring;

    fn index(&self, coord: Coordinate) -> &Self::Output {
        &self.data[coord.x + 7 * coord.y]
    }
}

impl IndexMut<Coordinate> for Board {
    fn index_mut(&mut self, coord: Coordinate) -> &mut Self::Output {
        &mut self.data[coord.x + 7 * coord.y]
    }
}

impl Debug for Board {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "")?;
        writeln!(f, "{:?}", &self.data[7 * 6..7 * 6 + 7])?;
        writeln!(f, "  {:?}", &self.data[7 * 5..7 * 5 + 7])?;
        writeln!(f, "    {:?}", &self.data[7 * 4..7 * 4 + 7])?;
        writeln!(f, "      {:?}", &self.data[7 * 3..7 * 3 + 7])?;
        writeln!(f, "        {:?}", &self.data[7 * 2..7 * 2 + 7])?;
        writeln!(f, "          {:?}", &self.data[7 * 1..7 * 1 + 7])?;
        writeln!(f, "            {:?}", &self.data[7 * 0..7 * 0 + 7])?;

        Ok(())
    }
}

File: src/Board.zig
const std = @import("std");

const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const Coordinate = @import("./Coordinate.zig");
const CoordinateIterator = @import("./Coordinate.zig").CoordinateIterator;

const assert = std.debug.assert;
const print = std.debug.print;

pub const Ring = union(enum) {
    empty,
    vacant,
    occupied: Marble,

    pub fn format(
        self: @This(),
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt;
        _ = options;

        try switch (self) {
            .empty => writer.print(".", .{}),
            .vacant => writer.print("O", .{}),
            .occupied => |marble| switch (marble) {
                .white => writer.print("\x1b[107;30m@\x1b[0m", .{}),
                .gray => writer.print("\x1b[100;30m@\x1b[0m", .{}),
                .black => writer.print("\x1b[97;40m@\x1b[0m", .{}),
            },
        };
    }
};

pub const Marble = enum(u2) {
    white,
    gray,
    black,
};

// Board fields
data: ArrayList(Ring),
// END fields

pub const Board = @This();
const Self = @This();

pub fn init(allocator: Allocator) !Self {
    var data = try ArrayList(Ring).initCapacity(allocator, 49);
    errdefer data.deinit();

    try data.appendNTimes(Ring.vacant, 49);

    var output = Self{ .data = data };
    var iter = CoordinateIterator.new(Coordinate.new(0, 0), Coordinate.new(6, 6), 6);

    while (iter.next()) |coord| {
        if (coord.x > coord.y + 3 or coord.y > coord.x + 3) {
            output.get_mut_raw(coord).* = Ring.empty;
        }
    }

    return output;
}

pub fn deinit(self: Self) void {
    self.data.deinit();
}

pub fn get_raw(self: *const Self, coord: Coordinate) Ring {
    assert(coord.x < 7 and coord.y < 7);
    return self.data.items[coord.x + 7 * coord.y];
}

pub fn get_mut_raw(self: *Self, coord: Coordinate) *Ring {
    assert(coord.x < 7 and coord.y < 7);
    return &self.data.items[coord.x + 7 * coord.y];
}

pub fn get(self: *const Self, coord: Coordinate) ?Ring {
    if (coord.x >= 7 or coord.y >= 7) {
        return null;
    }
    return self.data.items[coord.x + 7 * coord.y];
}

pub fn get_mut(self: *Self, coord: Coordinate) ?*Ring {
    if (coord.x >= 7 or coord.y >= 7) {
        return null;
    }
    return &self.data.items[coord.x + 7 * coord.y];
}

pub fn get_option(self: *const Self, coord: ?Coordinate) ?Ring {
    if (coord == null) {
        return null;
    }
    const coord_raw = coord.?;

    if (coord_raw.x >= 7 or coord_raw.y >= 7) {
        return null;
    }
    return self.data.items[coord_raw.x + 7 * coord_raw.y];
}

pub fn get_mut_option(self: *Self, coord: ?Coordinate) ?*Ring {
    if (coord == null) {
        return null;
    }
    const coord_raw = coord.?;

    if (coord_raw.x >= 7 or coord_raw.y >= 7) {
        return null;
    }
    return &self.data.items[coord_raw.x + 7 * coord_raw.y];
}

pub fn format(
    self: Self,
    comptime fmt: []const u8,
    options: std.fmt.FormatOptions,
    writer: anytype,
) !void {
    _ = fmt;
    _ = options;

    const loop_idx = [_]comptime_int{ 6, 5, 4, 3, 2, 1, 0 };
    comptime var i: comptime_int = undefined;
    inline for (loop_idx) |idx| {
        try writer.print("\n {}  " ++ " " ** (6 - idx), .{idx});
        i = 7 * idx;
        inline while (i < 7 * idx + 7) : (i += 1) {
            try writer.print("{} ", .{self.data.items[i]});
        }
    }
    try writer.print("\n\n            ", .{});

    i = 0;
    inline while (i < 7) : (i += 1) {
        try writer.print("{} ", .{i});
    }

    try writer.print("\n", .{});
}
